ktor-client/ktor-client-core/jvm/src/io/ktor/client/plugins/cache/storage/FileCacheStorage.kt

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.plugins.cache.storage
+
+import io.ktor.http.*
+import io.ktor.util.*
+import io.ktor.util.collections.*
+import io.ktor.util.date.*
+import io.ktor.utils.io.*
+import io.ktor.utils.io.core.*
+import io.ktor.utils.io.jvm.javaio.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.sync.*
+import java.io.*
+
+public class FileCacheStorage(
+    private val directory: File,
+    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
+) : CacheStorage {
+
+    private val mutexes = ConcurrentMap<String, Mutex>()
+
+    init {
+        directory.mkdirs()
+    }
+
+    override suspend fun store(url: Url, data: CachedResponseData): Unit = withContext(dispatcher) {
+        val urlHex = key(url)
+        val caches = readCache(urlHex).filterNot { it.varyKeys == data.varyKeys } + data
+        writeCache(urlHex, caches)
+    }
+
+    override suspend fun findAll(url: Url): Set<CachedResponseData> {
+        return readCache(key(url)).toSet()
+    }
+
+    override suspend fun find(url: Url, varyKeys: Map<String, String>): CachedResponseData? {
+        return readCache(key(url)).find { it.varyKeys == varyKeys }
+    }
+
+    private fun key(url: Url) = hex(url.toString().encodeToByteArray())
+
+    private suspend fun writeCache(urlHex: String, caches: List<CachedResponseData>) = coroutineScope {
+        val mutex = mutexes.computeIfAbsent(urlHex) { Mutex() }
+        mutex.withLock {
+            val channel = ByteChannel()
+            File(directory, urlHex).outputStream().buffered().use { output ->
+                launch {
```

Comment:
The file could be closed before the launch is finished if the `launch` block suspends. Could you please add a test with a large body and move the file `use` inside?

Reply:
`channel.copyTo(output)` is outside of `launch` block and it should ensure that file should be open as long as the channel is open. The channel is closed at the end of `launch` block. Do I miss something here?